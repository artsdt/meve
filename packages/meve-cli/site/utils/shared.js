import Vue from 'vue'

export const isArray = val => Array.isArray(val)

export function getMenuName(path) {
  return path.slice(path.lastIndexOf('/') + 1)
}

export function getLang(path) {
  return path.slice(1, path.lastIndexOf('/'))
}

export function buildTreeRelation(tree, parent, seen) {
  const flatTree = []

  tree.forEach((node) => {
    seen ? seen.push(node) : flatTree.push(node)
    node.parent = parent
    node.leaf = false

    if (isArray(node.children)) {
      buildTreeRelation(node.children, node, flatTree)
    } else {
      node.leaf = true
    }
  })

  return {
    tree,
    flatTree
  }
}

export function findNodeRelationChain(node) {
  const chain = []
  let parent = node.parent

  while (parent) {
    chain.unshift(parent)
    parent = parent.parent
  }

  return chain
}

export function nextTick() {
  return new Promise((resolve) => Vue.nextTick(resolve))
}

export function doubleRaf() {
  return new Promise((resolve) => {
    window.requestAnimationFrame(() => {
      window.requestAnimationFrame(resolve)
    })
  })
}

