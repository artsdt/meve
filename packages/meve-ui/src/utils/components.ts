import Vue, { ComponentOptions } from 'vue'
import { CombinedVueInstance } from 'vue/types/vue'

export interface MountComponentApi {
  instance: CombinedVueInstance<any, any, any, any, any>;
  unmount(): void;
}

export function mountComponent(component: ComponentOptions<Vue>, container: string, options = {}): MountComponentApi {
  const instance = new (Vue.extend(component))(options)
  const el = instance.$mount().$el
  const wrapper = document.querySelector(container)!
  wrapper.appendChild(el)

  return {
    instance,
    unmount() {
      instance.$destroy()
      wrapper.removeChild(el)
    },
  }
}

export function nextTick(): Promise<void> {
  return new Promise((resolve) => Vue.nextTick(resolve))
}
