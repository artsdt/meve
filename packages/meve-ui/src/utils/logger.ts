export function warn(message: string, from?: string) {
  if (process?.env?.NODE_ENV === 'development') {
    console.warn(`Warning: ${message} ${from ? `from ${from}` : ''}`)
  }
}
