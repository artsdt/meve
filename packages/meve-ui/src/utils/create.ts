import type Vue from 'vue'
import { VNodeData, ComponentOptions, VueConstructor, CreateElement, VNode, RenderContext } from 'vue'
import { PropsDefinition } from 'vue/types/options'
import { bigCamelize, isArray, isFunction, kebabCase } from './shared'
import { SlotsMixin } from './mixins/slots'

export type DefaultProps = Record<string, any>

export type DefaultSlots = {
  default?: ScopedSlot;
}

export type ScopedSlot<Props = any> = (props?: Props) => VNode[] | undefined

export type ScopedSlots = {
  [key: string]: ScopedSlot | undefined;
} & DefaultSlots

export type FunctionComponent<Props = DefaultProps, PropsDefinitions = PropsDefinition<Props>> = {
  (h: CreateElement, props: Props, slots: ScopedSlots, context: RenderContext<Props>): VNode | undefined;
  name?: string;
  props?: PropsDefinitions;
}

export type FunctionComponentPlugin = FunctionComponent & ComponentPlugin

export type TsxBaseProps<Slots> = {
  key: string | number;
  props: any;
  class: any;
  style: string | object[] | object;
  scopedSlots: Slots;
}

export type TsxComponent<Props, Events, Slots> = (props: Partial<Props & Events & TsxBaseProps<Slots>>) => VNode

export interface ComponentPlugin {
  install?: (app: VueConstructor, options?: any) => void;
}

export type ComponentOptionsPlugin = ComponentOptions<Vue> & ComponentPlugin

export function mergeSlots(context: RenderContext) {
  const scopedSlots = context.scopedSlots ?? context.data.scopedSlots ?? {}
  const slots = context.slots()

  Object.keys(slots).forEach((key) => {
    if (!scopedSlots[key]) {
      scopedSlots[key] = () => slots[key]
    }
  })

  return scopedSlots
}

export function wrapFunctionComponent(
  functionComponent: FunctionComponent
): ComponentOptionsPlugin & { functional?: boolean } {
  return {
    functional: true,
    props: functionComponent.props,
    render: (h: CreateElement, context: RenderContext): any =>
      functionComponent(h, context.props, mergeSlots(context), context),
  }
}

export function install(this: ComponentOptions<Vue>, Vue: VueConstructor) {
  Vue.component(this.name as string, this)
}

export function createComponent<Props = DefaultProps, Events = {}, Slots = {}>(
  name: string,
  sfc: ComponentOptionsPlugin | FunctionComponentPlugin
) {
  if (isFunction(sfc)) {
    sfc = wrapFunctionComponent(sfc as FunctionComponentPlugin)
  } else if ((sfc as ComponentOptionsPlugin).mixins) {
    ;(sfc as ComponentOptionsPlugin).mixins!.push(SlotsMixin)
  } else {
    ;(sfc as ComponentOptionsPlugin).mixins = [SlotsMixin]
  }

  sfc.name = `M${bigCamelize(name)}`
  sfc.install = install

  return sfc as TsxComponent<Props, Events, Slots>
}

export function createNamespace(name: string) {
  return {
    createComponent<Props, Events, Slots>(sfc: ComponentOptionsPlugin | FunctionComponentPlugin) {
      return createComponent<Props, Events, Slots>(name, sfc)
    },
    namespace: (suffix = '') => `m-${kebabCase(name)}${suffix}`,
  }
}

export type Context = RenderContext & { data: VNodeData & Record<string, any> }

export type InheritContext = Partial<VNodeData> & Record<string, any>

const inheritKeys = [
  'ref',
  'key',
  'style',
  'class',
  'attrs',
  'refInFor',
  'nativeOn',
  'directives',
  'staticClass',
  'staticStyle',
]

export function inherit(context: Context): InheritContext {
  return inheritKeys.reduce((ret, key) => {
    if (context.data[key]) {
      ret[key === 'nativeOn' ? 'on' : key] = context.data[key]
    }
    return ret
  }, {} as InheritContext)
}

export function emit(context: RenderContext, eventName: string, ...args: any[]) {
  const listeners = context.listeners[eventName]
  if (isArray(listeners)) {
    listeners.forEach((listener) => listener(...args))
  } else {
    listeners?.(...args)
  }
}
