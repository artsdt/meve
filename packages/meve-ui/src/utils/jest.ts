import Vue from 'vue'
import { Wrapper, config } from '@vue/test-utils'
import { isPlainObject } from './shared'

export const delay = (time = 0) => new Promise((resolve) => setTimeout(resolve, time))

export function mockOffset({
  offsetWidth,
  offsetHeight,
  clientWidth,
  clientHeight,
  offsetLeft,
  offsetTop,
  scrollWidth,
  scrollHeight,
}: {
  offsetWidth?: number;
  offsetHeight?: number;
  clientWidth?: number;
  clientHeight?: number;
  offsetLeft?: number;
  offsetTop?: number;
  scrollWidth?: number;
  scrollHeight?: number;
} = {}) {
  Object.defineProperties(HTMLElement.prototype, {
    offsetWidth: {
      get() {
        return offsetWidth ?? (parseFloat(window.getComputedStyle(this).width) || 0)
      },
    },
    offsetHeight: {
      get() {
        return offsetHeight ?? (parseFloat(window.getComputedStyle(this).height) || 0)
      },
    },
    clientWidth: {
      get() {
        return clientWidth ?? (parseFloat(window.getComputedStyle(this).width) || 0)
      },
    },
    clientHeight: {
      get() {
        return clientHeight ?? (parseFloat(window.getComputedStyle(this).height) || 0)
      },
    },
    offsetLeft: {
      get() {
        return offsetLeft ?? (parseFloat(window.getComputedStyle(this).marginLeft) || 0)
      },
    },
    offsetTop: {
      get() {
        return offsetTop ?? (parseFloat(window.getComputedStyle(this).marginTop) || 0)
      },
    },
    offsetParent: {
      get() {
        return this.parentNode ?? {}
      },
    },
    scrollWidth: {
      get() {
        return scrollWidth ?? 0
      },
    },
    scrollHeight: {
      get() {
        return scrollHeight ?? 0
      },
    },
  })

  return {
    mockRestore() {
      Object.defineProperties(HTMLElement.prototype, {
        offsetWidth: {
          get() {
            return parseFloat(window.getComputedStyle(this).width) || 0
          },
        },
        offsetHeight: {
          get() {
            return parseFloat(window.getComputedStyle(this).height) || 0
          },
        },
        clientWidth: {
          get() {
            return parseFloat(window.getComputedStyle(this).width) || 0
          },
        },
        clientHeight: {
          get() {
            return parseFloat(window.getComputedStyle(this).height) || 0
          },
        },
        offsetLeft: {
          get() {
            return parseFloat(window.getComputedStyle(this).marginLeft) || 0
          },
        },
        offsetTop: {
          get() {
            return parseFloat(window.getComputedStyle(this).marginTop) || 0
          },
        },
        offsetParent: {
          get() {
            return this.parentNode ?? {}
          },
        },
        scrollWidth: {
          get() {
            return 0
          },
        },
        scrollHeight: {
          get() {
            return 0
          },
        },
      })
    },
  }
}

export async function triggerDrag(
  el: Wrapper<Vue> | HTMLElement | Window,
  x = 0,
  y = 0,
  triggerTouchend = true,
  triggerDocument = false
) {
  const delegationEl = triggerDocument ? document : el

  await trigger(el, 'touchstart', 0, 0)
  await trigger(delegationEl, 'touchmove', x / 4, y / 4)
  await trigger(delegationEl, 'touchmove', x / 3, y / 3)
  await trigger(delegationEl, 'touchmove', x / 2, y / 2)
  await trigger(delegationEl, 'touchmove', x, y)

  if (triggerTouchend) {
    await trigger(delegationEl, 'touchend', x, y)
  }
}

export function getTouch(el: Element | Window | Document, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5,
  }
}

export function trigger(
  wrapper: Wrapper<Vue> | HTMLElement | Window | Document,
  eventName: string,
  x = 0,
  y = 0,
  deltaX = 0,
  deltaY = 0,
  offsetX = 0,
  offsetY = 0,
  args = {}
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper
  const touchList = [getTouch(el, x, y)]

  const event = document.createEvent('CustomEvent')
  event.initCustomEvent(eventName, true, true, {})

  Object.assign(event, {
    clientX: x,
    clientY: y,
    deltaX,
    deltaY,
    offsetX,
    offsetY,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList,
    ...args,
  })

  el.dispatchEvent(event)

  return Vue.nextTick()
}

export function mockStubs() {
  const originStubs = config.stubs

  config.stubs = {}

  return {
    mockRestore() {
      config.stubs = originStubs
    },
  }
}

export function mockDoubleRaf() {
  const originMethod = window.requestAnimationFrame

  Object.assign(window, {
    requestAnimationFrame(fn: any) {
      setTimeout(fn, 16)
    },
  })

  return {
    mockRestore() {
      window.requestAnimationFrame = originMethod
    },
  }
}

export function mockScrollTo(Element: any) {
  Element.prototype.scrollTo = function (x: ScrollToOptions | number, y?: number) {
    if (isPlainObject(x)) {
      this.scrollLeft = x.left
      this.scrollTop = x.top
    } else {
      this.scrollLeft = x
      this.scrollTop = y
    }
  }
}
