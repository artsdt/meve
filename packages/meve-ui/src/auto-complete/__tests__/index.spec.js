import AutoComplete from '..'
import { mount } from '@vue/test-utils'
import { delay, mockOffset, mockStubs, trigger } from '../../utils/jest'

mockOffset()

test('test auto-complete value binding', () => {
  const wrapper = mount(AutoComplete, {
    propsData: {
      value: 'hello',
    },
  })

  expect(wrapper.html()).toMatchSnapshot()
  wrapper.find('input').setValue('hello world')
  wrapper.find('input').trigger('change')
  expect(wrapper.emitted().input.length).toBe(1)
  expect(wrapper.emitted().change.length).toBe(1)
  expect(wrapper.emitted().input[0][0]).toBe('hello world')
  expect(wrapper.emitted().change[0][0]).toBe('hello world')
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.destroy()
})

test('test auto-complete size', () => {
  const wrapper = mount({
    components: {
      [AutoComplete.name]: AutoComplete,
    },
    template: `
      <div>
        <m-auto-complete/>
        <m-auto-complete size="small"/>
        <m-auto-complete size="mini"/>
      </div>
    `,
  })

  expect(wrapper.html()).toMatchSnapshot()

  wrapper.destroy()
})

test('test auto-complete focus & blur', async () => {
  const wrapper = mount(
    AutoComplete,
    {
      propsData: {
        value: 'hello',
        options: [
          {
            label: 'item 1',
            value: 'item 1',
          },
        ],
      },
    },
    { attachTo: document.body }
  )

  wrapper.vm.focus()
  await trigger(wrapper.find('.m-input__input'), 'focus')
  await delay(300)
  expect(wrapper.html()).toMatchSnapshot()
  expect(wrapper.emitted().focus.length).toBe(1)

  wrapper.vm.blur()
  await trigger(wrapper.find('.m-input__input'), 'blur')
  await delay(300)
  expect(wrapper.html()).toMatchSnapshot()
  expect(wrapper.emitted().blur.length).toBe(1)
  expect(document.querySelector('.m-popover__container')).toBeTruthy()
  wrapper.destroy()
  document.body.innerHTML = ''
})

test('test auto-complete textarea focus & blur', async () => {
  const wrapper = mount(AutoComplete, { propsData: { textarea: true } })

  wrapper.vm.focus()
  await trigger(wrapper.find('.m-input__textarea'), 'focus')
  expect(wrapper.html()).toMatchSnapshot()
  expect(wrapper.emitted().focus.length).toBe(1)

  wrapper.vm.blur()
  await trigger(wrapper.find('.m-input__textarea'), 'blur')
  expect(wrapper.html()).toMatchSnapshot()
  expect(wrapper.emitted().blur.length).toBe(1)

  wrapper.destroy()
})

test('test auto-complete maxlength', async () => {
  const wrapper = mount(AutoComplete, { propsData: { maxlength: 6, value: '' } })

  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.setProps({ value: 'hello' })

  expect(wrapper.html()).toMatchSnapshot()

  wrapper.destroy()
})

test('test auto-complete click event', async () => {
  const wrapper = mount(AutoComplete)

  await wrapper.find('.m-input').trigger('click')

  expect(wrapper.emitted().click.length).toBe(1)
  wrapper.destroy()
})

test('test auto-complete clear', async () => {
  const wrapper = mount(AutoComplete, { propsData: { clearable: true, value: 'hello' } })

  expect(wrapper.html()).toMatchSnapshot()
  expect(wrapper.find('.m-input__clear-icon').classes().includes('m-input--clear-icon-visible')).toBeTruthy()

  await wrapper.find('.m-input__clear-icon').trigger('click')
  expect(wrapper.emitted().clear.length).toBe(1)
  expect(wrapper.emitted().input.length).toBe(1)
  expect(wrapper.emitted().input[0][0]).toBe(undefined)

  await wrapper.setProps({ value: '' })
  expect(wrapper.find('.m-input__clear-icon').classes().includes('m-input--clear-icon-visible')).toBeFalsy()
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.destroy()
})

test('test auto-complete slots', () => {
  const wrapper = mount(AutoComplete, {
    scopedSlots: {
      'prepend-icon': '<div>prepend icon</div>',
      'append-icon': '<div>append icon</div>',
    },
  })

  expect(wrapper.html()).toMatchSnapshot()

  wrapper.destroy()
})

test('test auto-complete disabled', async () => {
  const wrapper = mount(AutoComplete, {
    propsData: {
      disabled: true,
      clearable: true,
    },
  })

  await wrapper.find('.m-input').trigger('click')
  await wrapper.find('.m-input__clear-icon').trigger('click')
  expect(wrapper.emitted().click).toBeFalsy()
  expect(wrapper.emitted().clear).toBeFalsy()

  wrapper.destroy()
})

test('test auto-complete readonly', async () => {
  const wrapper = mount(AutoComplete, {
    propsData: {
      value: 'hello',
      readonly: true,
      clearable: true,
    },
  })

  await wrapper.find('.m-input').trigger('click')
  await wrapper.find('.m-input__clear-icon').trigger('click')
  expect(wrapper.emitted().click.length).toBe(2)
  expect(wrapper.emitted().clear).toBeFalsy()

  wrapper.destroy()
})

test('test auto-complete validation', async () => {
  const { mockRestore } = mockStubs()
  const wrapper = mount(AutoComplete, {
    propsData: {
      value: '',
      rules: [(v) => (v && v.length > 6) || '长度必须大于6'],
    },
  })

  wrapper.vm.validate()
  await delay(100)
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.vm.resetValidation()
  await delay(100)
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.vm.reset()
  expect(wrapper.emitted().input.length).toBe(1)
  expect(wrapper.emitted().input[0][0]).toBe(undefined)
  mockRestore()

  wrapper.destroy()
})

test('test auto-complete popover click', async () => {
  const wrapper = mount(
    AutoComplete,
    {
      propsData: {
        value: '',
        options: [
          {
            label: 'item 1',
            value: 'item 1',
          },
        ],
      },
    },
    { attachTo: document.body }
  )

  wrapper.find('input').setValue('h')

  await delay(300)

  await trigger(document.querySelector('.m-auto-complete__option'), 'click')
  expect(wrapper.emitted().input.length).toBe(2)
  expect(wrapper.emitted().input[1][0]).toBe('item 1')
  expect(wrapper.emitted().select.length).toBe(1)
  expect(wrapper.emitted().select[0][0]).toBe('item 1')

  wrapper.destroy()
  document.body.innerHTML = ''
})

test('test auto-complete popover keyboard', async () => {
  const wrapper = mount(
    AutoComplete,
    {
      propsData: {
        value: '',
        options: [
          {
            label: 'item 1',
            value: 'item 1',
          },
          {
            label: 'item 2',
            value: 'item 2',
          },
        ],
      },
    },
    { attachTo: document.body }
  )

  wrapper.find('input').setValue('h')

  await delay(300)

  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'ArrowDown' })

  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'Enter' })
  expect(wrapper.emitted().input.length).toBe(2)
  expect(wrapper.emitted().input[1][0]).toBe('item 2')

  wrapper.destroy()
  document.body.innerHTML = ''
})

test('test auto-complete popover open and close', async () => {
  const { mockRestore } = mockStubs()
  const wrapper = mount(
    AutoComplete,
    {
      propsData: {
        value: '',
        options: [
          {
            label: 'item 1',
            value: 'item 1',
          },
        ],
      },
    },
    { attachTo: document.body }
  )

  expect(document.querySelector('.m-popover__container').style.display).toBe('none')

  wrapper.vm.open()
  await delay(300)
  expect(document.querySelector('.m-popover__container').style.display).toBe('')
  await trigger(wrapper.find('.m-input__input'), 'blur')
  await delay(300)

  expect(document.querySelector('.m-popover__container').style.display).toBe('none')

  wrapper.destroy()
  document.body.innerHTML = ''
  mockRestore()
})

test('test auto-complete mouse enter options cant close', async () => {
  const { mockRestore } = mockStubs()
  const wrapper = mount(
    AutoComplete,
    {
      propsData: {
        value: '',
        options: [
          {
            label: 'item 1',
            value: 'item 1',
          },
        ],
      },
    },
    { attachTo: document.body }
  )

  expect(document.querySelector('.m-popover__container').style.display).toBe('none')

  wrapper.vm.open()
  await delay(300)
  expect(document.querySelector('.m-popover__container').style.display).toBe('')
  await trigger(document.querySelector('.m-auto-complete__options'), 'mouseenter')
  await trigger(wrapper.find('.m-input__input'), 'blur')
  await delay(300)
  expect(document.querySelector('.m-popover__container').style.display).toBe('')

  await trigger(document.querySelector('.m-auto-complete__options'), 'mouseleave')
  await trigger(wrapper.find('.m-input__input'), 'blur')
  await delay(300)
  expect(document.querySelector('.m-popover__container').style.display).toBe('none')

  wrapper.destroy()
  document.body.innerHTML = ''
  mockRestore()
})

test('test auto-complete mouse enter option and enter', async () => {
  const wrapper = mount(
    AutoComplete,
    {
      propsData: {
        value: '',
        options: [
          {
            label: 'item 1',
            value: 'item 1',
          },
          {
            label: 'item 2',
            value: 'item 2',
          },
        ],
      },
    },
    { attachTo: document.body }
  )

  wrapper.find('input').setValue('h')

  await delay(300)

  await trigger(document.querySelector('.m-auto-complete__option'), 'mouseenter')

  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'Enter' })
  expect(wrapper.emitted().input.length).toBe(2)
  expect(wrapper.emitted().input[1][0]).toBe('item 1')

  wrapper.destroy()
  document.body.innerHTML = ''
})

test('test auto-complete esc close', async () => {
  const wrapper = mount(
    AutoComplete,
    {
      propsData: {
        value: '',
        options: [
          {
            label: 'item 1',
            value: 'item 1',
          },
          {
            label: 'item 2',
            value: 'item 2',
          },
        ],
      },
    },
    { attachTo: document.body }
  )

  wrapper.find('input').setValue('h')
  await delay(300)
  expect(document.querySelector('.m-popover__container').style.display).toBe('')
  await trigger(window, 'keydown', 0, 0, 0, 0, 0, 0, { code: 'Escape' })
  await delay(300)
  expect(document.querySelector('.m-popover__container').style.display).toBe('none')
  wrapper.destroy()
  document.body.innerHTML = ''
})
