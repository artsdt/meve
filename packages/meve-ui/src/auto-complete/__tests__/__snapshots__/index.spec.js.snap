// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test auto-complete clear 1`] = `
"<div class=\\"m-auto-complete\\">
  <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-input m--box m-input--normal\\"><input id=\\"m-input-8\\" autocomplete=\\"off\\" type=\\"text\\" class=\\"m-input__input\\"><i input-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-close-circle m-input__clear-icon m-input--clear-icon-visible\\"></i></div>
      <transition-stub name=\\"m-form-details\\"></transition-stub>
    </div>
    <div class=\\"m-teleport\\"></div>
  </div>
</div>"
`;

exports[`test auto-complete clear 2`] = `
"<div class=\\"m-auto-complete\\">
  <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-input m--box m-input--normal\\"><input id=\\"m-input-8\\" autocomplete=\\"off\\" type=\\"text\\" class=\\"m-input__input\\"><i input-cover=\\"true\\" class=\\"m-icon m--box meve-icon--set meve-icon-close-circle m-input__clear-icon\\"></i></div>
      <transition-stub name=\\"m-form-details\\"></transition-stub>
    </div>
    <div class=\\"m-teleport\\"></div>
  </div>
</div>"
`;

exports[`test auto-complete focus & blur 1`] = `
"<div class=\\"m-auto-complete\\">
  <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-input m--box m-input--normal m-input--outline\\"><input id=\\"m-input-4\\" autocomplete=\\"off\\" type=\\"text\\" class=\\"m-input__input\\"></div>
      <transition-stub name=\\"m-form-details\\"></transition-stub>
    </div>
    <div class=\\"m-teleport\\"></div>
  </div>
</div>"
`;

exports[`test auto-complete focus & blur 2`] = `
"<div class=\\"m-auto-complete\\">
  <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-input m--box m-input--normal\\"><input id=\\"m-input-4\\" autocomplete=\\"off\\" type=\\"text\\" class=\\"m-input__input\\"></div>
      <transition-stub name=\\"m-form-details\\"></transition-stub>
    </div>
    <div class=\\"m-teleport\\"></div>
  </div>
</div>"
`;

exports[`test auto-complete maxlength 1`] = `
"<div class=\\"m-auto-complete\\">
  <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-input m--box m-input--normal\\"><input id=\\"m-input-6\\" autocomplete=\\"off\\" type=\\"text\\" maxlength=\\"6\\" class=\\"m-input__input\\">
        <div class=\\"m-input__maxlength\\">0 / 6</div>
      </div>
      <transition-stub name=\\"m-form-details\\"></transition-stub>
    </div>
    <div class=\\"m-teleport\\"></div>
  </div>
</div>"
`;

exports[`test auto-complete maxlength 2`] = `
"<div class=\\"m-auto-complete\\">
  <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-input m--box m-input--normal\\"><input id=\\"m-input-6\\" autocomplete=\\"off\\" type=\\"text\\" maxlength=\\"6\\" class=\\"m-input__input\\">
        <div class=\\"m-input__maxlength\\">5 / 6</div>
      </div>
      <transition-stub name=\\"m-form-details\\"></transition-stub>
    </div>
    <div class=\\"m-teleport\\"></div>
  </div>
</div>"
`;

exports[`test auto-complete size 1`] = `
"<div>
  <div class=\\"m-auto-complete\\">
    <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
      <div class=\\"m-form-item\\">
        <div class=\\"m-input m--box m-input--normal\\"><input id=\\"m-input-1\\" autocomplete=\\"off\\" type=\\"text\\" class=\\"m-input__input\\"></div>
        <transition-stub name=\\"m-form-details\\"></transition-stub>
      </div>
      <div class=\\"m-teleport\\"></div>
    </div>
  </div>
  <div class=\\"m-auto-complete\\">
    <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
      <div class=\\"m-form-item\\">
        <div class=\\"m-input m--box m-input--small\\"><input id=\\"m-input-2\\" autocomplete=\\"off\\" type=\\"text\\" class=\\"m-input__input\\"></div>
        <transition-stub name=\\"m-form-details\\"></transition-stub>
      </div>
      <div class=\\"m-teleport\\"></div>
    </div>
  </div>
  <div class=\\"m-auto-complete\\">
    <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
      <div class=\\"m-form-item\\">
        <div class=\\"m-input m--box m-input--mini\\"><input id=\\"m-input-3\\" autocomplete=\\"off\\" type=\\"text\\" class=\\"m-input__input\\"></div>
        <transition-stub name=\\"m-form-details\\"></transition-stub>
      </div>
      <div class=\\"m-teleport\\"></div>
    </div>
  </div>
</div>"
`;

exports[`test auto-complete slots 1`] = `
"<div class=\\"m-auto-complete\\">
  <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-input m--box m-input--normal\\">
        <div>prepend icon</div><input id=\\"m-input-9\\" autocomplete=\\"off\\" type=\\"text\\" class=\\"m-input__input\\">
        <div>append icon</div>
      </div>
      <transition-stub name=\\"m-form-details\\"></transition-stub>
    </div>
    <div class=\\"m-teleport\\"></div>
  </div>
</div>"
`;

exports[`test auto-complete textarea focus & blur 1`] = `
"<div class=\\"m-auto-complete\\">
  <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-input m--box m-input--normal m-input--outline m-input--resizeable\\" style=\\"height: auto;\\"><textarea id=\\"m-input-5\\" autocomplete=\\"off\\" rows=\\"5\\" class=\\"m-input__textarea m-input--normal-textarea\\"></textarea></div>
      <transition-stub name=\\"m-form-details\\"></transition-stub>
    </div>
    <div class=\\"m-teleport\\"></div>
  </div>
</div>"
`;

exports[`test auto-complete textarea focus & blur 2`] = `
"<div class=\\"m-auto-complete\\">
  <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-input m--box m-input--normal m-input--resizeable\\" style=\\"height: auto;\\"><textarea id=\\"m-input-5\\" autocomplete=\\"off\\" rows=\\"5\\" class=\\"m-input__textarea m-input--normal-textarea\\"></textarea></div>
      <transition-stub name=\\"m-form-details\\"></transition-stub>
    </div>
    <div class=\\"m-teleport\\"></div>
  </div>
</div>"
`;

exports[`test auto-complete validation 1`] = `
"<div class=\\"m-auto-complete\\">
  <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-input m--box m-input--normal m-input--validation-error-border\\"><input id=\\"m-input-12\\" autocomplete=\\"off\\" type=\\"text\\" class=\\"m-input__input\\"></div>
      <div class=\\"m-form-details\\">长度必须大于6</div>
    </div>
    <div class=\\"m-teleport\\"></div>
  </div>
</div>"
`;

exports[`test auto-complete validation 2`] = `
"<div class=\\"m-auto-complete\\">
  <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-input m--box m-input--normal\\"><input id=\\"m-input-12\\" autocomplete=\\"off\\" type=\\"text\\" class=\\"m-input__input\\"></div>
      <!---->
    </div>
    <div class=\\"m-teleport\\"></div>
  </div>
</div>"
`;

exports[`test auto-complete value binding 1`] = `
"<div class=\\"m-auto-complete\\">
  <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-input m--box m-input--normal\\"><input id=\\"m-input-0\\" autocomplete=\\"off\\" type=\\"text\\" class=\\"m-input__input\\"></div>
      <transition-stub name=\\"m-form-details\\"></transition-stub>
    </div>
    <div class=\\"m-teleport\\"></div>
  </div>
</div>"
`;

exports[`test auto-complete value binding 2`] = `
"<div class=\\"m-auto-complete\\">
  <div class=\\"m-popover m-auto-complete__popover\\" auto-complete-cover=\\"true\\">
    <div class=\\"m-form-item\\">
      <div class=\\"m-input m--box m-input--normal\\"><input id=\\"m-input-0\\" autocomplete=\\"off\\" type=\\"text\\" class=\\"m-input__input\\"></div>
      <transition-stub name=\\"m-form-details\\"></transition-stub>
    </div>
    <div class=\\"m-teleport\\"></div>
  </div>
</div>"
`;
